### PR TITLE
#2908 Add Voided ContractStatus

### PR DESCRIPTION
**Pull request by cristinamghita**
#2908 Provide Doc Action voiding for contract period

### CODE REVIEW

File: de.metas.contracts/src/main/java/de/metas/contracts/flatrate/process/C_Flatrate_Term_Extend.java [hunk 1]
```diff
@@ -26,24 +27,29 @@
...
 import de.metas.contracts.model.X_C_Flatrate_Term;
+import de.metas.document.engine.IDocument;
 import de.metas.process.JavaProcess;
 import de.metas.process.Param;
 
 public class C_Flatrate_Term_Extend
 		extends JavaProcess
 {
+	final private static IQueryBL queryBL = Services.get(IQueryBL.class);
+	final private static IFlatrateBL flatrateBL = Services.get(IFlatrateBL.class);
```
Comment (teosarca) [inline de.metas.contracts/src/main/java/de/metas/contracts/flatrate/process/C_Flatrate_Term_Extend.java]: don't make those static pls

File: de.metas.contracts/src/main/java/de/metas/contracts/impl/ContractChangeBL.java [hunk 2]
```diff
@@ -86,31 +86,41 @@ public void cancelContract(@NonNull final I_C_Flatrate_Term currentTerm,
 			cancelContractIfNotCanceledAlready(initialContract, contractChangeParameters);
 		}
 	}
-	
+
+	private boolean isVoidSingleContract(final @NonNull ContractChangeParameters contractChangeParameters)
```
Comment (teosarca) [inline de.metas.contracts/src/main/java/de/metas/contracts/impl/ContractChangeBL.java]: why not having this method in ContractChangeParameters  class?

File: de.metas.contracts/src/main/java/de/metas/contracts/impl/ContractChangeBL.java [hunk 3]
```diff
@@ -86,31 +86,41 @@ public void cancelContract(@NonNull final I_C_Flatrate_Term currentTerm,
...
+	{
+		return ChangeTerm_ACTION_VoidSingleContract.equals(extractAction(contractChangeParameters));
+	}
+
+	private String extractAction(final @NonNull ContractChangeParameters contractChangeParameters)
```
Comment (teosarca) [inline de.metas.contracts/src/main/java/de/metas/contracts/impl/ContractChangeBL.java]: i think this is too much abstraction.... we don't need this method. More, it's used in only one place.

File: de.metas.contracts/src/main/java/de/metas/contracts/impl/ContractChangeBL.java [hunk 4]
```diff
@@ -86,31 +86,41 @@ public void cancelContract(@NonNull final I_C_Flatrate_Term currentTerm,
...
 		final I_C_Flatrate_Term ancestor = flatrateDAO.retrieveAncestorFlatrateTerm(currentTerm);
-		 if (ancestor != null && contractChangeParameters.isOnlyTerminateCurrentTerm())
-		 {
-			 ancestor.setC_FlatrateTerm_Next(null);
-			 ancestor.setAD_PInstance_EndOfTerm(null);
-			 setAncestorMasterEndDateWhenUnlinkContract(ancestor);
-			 
-			 InterfaceWrapperHelper.save(ancestor);
-		 }
+		if (ancestor != null && isVoidSingleContract(contractChangeParameters))
```
Comment (teosarca) [inline de.metas.contracts/src/main/java/de/metas/contracts/impl/ContractChangeBL.java]: optimization hint: call isVoidSingleContract(contractChangeParameters) before loading "ancestor".
If that method returns false you can return from this method directly (shortcut), without loading that ancestor.

File: de.metas.contracts/src/main/java/de/metas/contracts/flatrate/process/C_Flatrate_Term_Extend.java (hunk 1 above)
Comment (cristinamghita) [reply de.metas.contracts/src/main/java/de/metas/contracts/flatrate/process/C_Flatrate_Term_Extend.java]: Why shall not be static?

File: de.metas.contracts/src/main/java/de/metas/contracts/impl/ContractChangeBL.java (hunk 4 above)
Comment (cristinamghita) [reply de.metas.contracts/src/main/java/de/metas/contracts/impl/ContractChangeBL.java]: thx!

File: de.metas.contracts/src/main/java/de/metas/contracts/flatrate/process/C_Flatrate_Term_Extend.java [hunk 5]
```diff
@@ -71,27 +75,23 @@ protected String doIt() throws Exception
...
-					.setClient_ID() // .setApplyAccessRules requires a more sophisticated role access setup which we urrently don't have to the user "system" () and this is not a "user"
-					// process
+			final Iterator<I_C_Flatrate_Term> termsToExtend = queryBL.createQueryBuilder(I_C_Flatrate_Term.class)
+					.addInArrayFilter(I_C_Flatrate_Term.COLUMNNAME_AD_PInstance_EndOfTerm_ID, 0, null)
+					.addEqualsFilter(I_C_Flatrate_Term.COLUMN_DocStatus, IDocument.STATUS_Completed)
+					.addCompareFilter(I_C_Flatrate_Term.COLUMN_NoticeDate, Operator.LESS, SystemTime.asTimestamp())
+					.addNotInArrayFilter(I_C_Flatrate_Term.COLUMN_ContractStatus, Arrays.asList(X_C_Flatrate_Term.CONTRACTSTATUS_Quit, X_C_Flatrate_Term.CONTRACTSTATUS_Voided))
+					.orderBy().addColumn(I_C_Flatrate_Term.COLUMN_C_Flatrate_Term_ID).endOrderBy()
+					.create()
+					.setClient_ID()
```
Comment (teosarca) [inline de.metas.contracts/src/main/java/de/metas/contracts/flatrate/process/C_Flatrate_Term_Extend.java]: not now... next time pls use org.adempiere.ad.dao.IQueryBuilder.addOnlyContextClient()

File: de.metas.contracts/src/main/java/de/metas/contracts/impl/ContractChangeBL.java [hunk 6]
```diff
@@ -86,31 +86,33 @@ public void cancelContract(@NonNull final I_C_Flatrate_Term currentTerm,
...
-		final I_C_Flatrate_Term ancestor = flatrateDAO.retrieveAncestorFlatrateTerm(currentTerm);
-		 if (ancestor != null && contractChangeParameters.isOnlyTerminateCurrentTerm())
-		 {
-			 ancestor.setC_FlatrateTerm_Next(null);
-			 ancestor.setAD_PInstance_EndOfTerm(null);
-			 setAncestorMasterEndDateWhenUnlinkContract(ancestor);
-			 
-			 InterfaceWrapperHelper.save(ancestor);
-		 }
+		if (contractChangeParameters.isVoidSingleContract())
```
Comment (teosarca) [inline de.metas.contracts/src/main/java/de/metas/contracts/impl/ContractChangeBL.java]: why not?
```
if (!contractChangeParameters.isVoidSingleContract())
{
   return;
}
```

**Pull request merged**